### PR TITLE
dissect: save image path to LoopDevice and use it when dissect loop device

### PR DESCRIPTION
**Pull request by yuwata**
Follow-up for #24322.

### CODE REVIEW

File: src/shared/blockdev-util.c [hunk 1]
```diff
@@ -509,19 +509,26 @@ int block_device_resize_partition(
...
-                return r;
+        if (!dev) {
+                struct stat st;
+
+                if (fstat(fd, &st) < 0)
+                        return -errno;
+
+                r = sd_device_new_from_stat_rdev(&dev_unref, &st);
+                if (r < 0)
+                        return r;
```
Comment (poettering) [inline src/shared/blockdev-util.c:575]: maybe time for a sd_device_new_from_fd()?

File: src/shared/loop-util.c [hunk 2]
```diff
@@ -332,6 +333,16 @@ static int loop_device_make_internal(
                         return r;
         }
 
+        if (path) {
+                backing_file = strdup(path);
+                if (!backing_file)
+                        return -ENOMEM;
```
Comment (poettering) [inline src/shared/loop-util.c]: maybe use path_make_absolute_cwd() and path_simplify()?

File: src/shared/dissect-image.c [hunk 3]
```diff
@@ -191,14 +191,10 @@ static int make_partition_devname(
 }
 #endif
 
-int dissect_image(
-                int fd,
-                const char *image_path,
+int dissect_loop_device(
```
Comment (poettering) [inline src/shared/dissect-image.c]: as i mentoined in that other PR: I am not convinced this is desirable: with this change we can only dissect stuff with privileges. Before we could analyze regular files too, without loopback devices, without privileges. I'd really like to keep that, and make this something we make use of in tests...

**Comment by poettering** [comment]
looks mostly good, but I am not convinced we should go `LoopDevice` only for the dissection code: being able to dissect regular files and thus operate without privs is a good thing we should keep and make use of in tests at the very least.

File: src/shared/dissect-image.c (hunk 3 above)
Comment (yuwata) [reply src/shared/dissect-image.c]: Oh, was it possible to call this without privileges? I did not know that.

Comment (yuwata) [reply src/shared/dissect-image.c]: At least on the current git HEAD, `dissect_image()` does not work for file descriptor for a regular file. Also, not work for block devices without privilege.
I guess we can modify the logic to support regular file or without privilege, but l'd like to do that later.

File: src/shared/blockdev-util.c (hunk 1 above)
Comment (yuwata) [reply src/shared/blockdev-util.c:575]: Yeah, but I'd like to do that in another later PR.

**Comment by yuwata** [comment]
@poettering Thank you for the comments. Again, dissect_image() is revived in the current version. PTAL.

File: src/shared/blockdev-util.c [hunk 4]
```diff
@@ -539,6 +546,12 @@ int block_device_remove_all_partitions(int fd) {
         if (r < 0)
                 return r;
 
+        if (fd < 0) {
+                fd = sd_device_open(dev, O_CLOEXEC|O_NONBLOCK|O_NOCTTY|O_RDONLY);
+                if (fd < 0)
+                        return fd;
+        }
```
Comment (poettering) [inline src/shared/blockdev-util.c:590]: you are leaking the fd here

Comment (yuwata) [reply src/shared/blockdev-util.c:590]: Ouch. Fixed.

**Pull request merged**